libs/lb/lb-rs/src/text/buffer.rs

```diff
@@ -0,0 +1,583 @@
+    }
+
+    fn invert(&self, op: &Operation) -> InverseOperation {
+        let mut inverse = InverseOperation { replace: None, select: self.selection };
+        if let Operation::Replace(replace) = op {
+            inverse.replace = Some(self.invert_replace(replace));
+        }
+        inverse
+    }
+
+    fn invert_replace(&self, replace: &Replace) -> Replace {
+        let Replace { range, text } = replace;
+        let byte_range = self.segs.range_to_byte(*range);
+        let replaced_text = self[byte_range].into();
+        let replacement_range = (range.start(), range.start() + text.len());
+        Replace { range: replacement_range, text: replaced_text }
+    }
+}
+
+#[derive(Default)]
+struct Ops {
+    /// Operations that have been received by the buffer
+    all: Vec<Operation>,
+    meta: Vec<OpMeta>,
+
+    /// Sequence number of the first unapplied operation. Operations starting with this one are queued for processing.
+    processed_seq: usize,
+
+    /// Operations that have been applied to the buffer and already transformed, in order of application. Each of these
+    /// operations is based on the previous operation in this list, with the first based on the history base. Derived
+    /// from other data and invalidated by some undo/redo flows.
+    transformed: Vec<Operation>,
+
+    /// Operations representing the inverse of the operations in `transformed_ops`, in order of application. Useful for
+    /// undoing operations. The data model differs because an operation that replaces text containing the cursor needs
+    /// two operations to revert the text and cursor. Derived from other data and invalidated by some undo/redo flows.
+    transformed_inverted: Vec<InverseOperation>,
+}
+
+impl Ops {
+    fn len(&self) -> usize {
+        self.all.len()
+    }
+
+    fn is_undo_checkpoint(&self, idx: usize) -> bool {
+        if idx == 0 {
+            return true;
+        }
+        if idx == self.len() {
+            return true;
+        }
+
+        let meta = &self.meta[idx];
+        let prev_meta = &self.meta[idx - 1];
+        if meta.timestamp - prev_meta.timestamp > Duration::from_millis(500) {
+            return true;
+        }
+
+        let op = &self.all[idx];
+        if meta.base != prev_meta.base && matches!(op, Operation::Select(..)) {
+            return true;
+        }
+
+        false
+    }
+}
+
+#[derive(Default)]
+struct External {
+    /// Text last loaded into the editor. Used as a reference point for merging out-of-editor changes with in-editor
+    /// changes, similar to a base in a 3-way merge. May be a state that never appears in the buffer's history.
+    text: String,
+
+    /// Index of the last external operation referenced when merging changes. May be ahead of current.seq if there has
+    /// not been a call to `update()` (updates current.seq) since the last call to `reload()` (assigns new greatest seq
+    /// to `external_seq`).
```

Comment:
workspace calls `editor.show()` and therefore `update()` referenced here after the only place it calls reload so I think we're good